33-green-wrasse/technical_specification.md

```diff
@@ -0,0 +1,340 @@
+- `access_starts`: data (when the access permission should start)
+- `access_ends`: date (when the access permission should end)
+- `request_created`: date (when the request has been created)
+- `status`: enum (allowed/denied/pending)
+- `status_changed`: date (when the status was last changed)
+- `changed_by`: string (user ID of the data steward)
+
+The `status` field should be automatically set to "pending" upon creation,
+and the `request_created` field should be filled with the current date.
+
+The `email` address is stored in order to notify the user. This address can
+be different from the registered user address. The `full_user_name` is stored
+so that it can be displayed in the Access Request Browser without requesting
+it from the User Registry.
+
+Again, note that we maintain only one pair of dates to specify the validity
+period, not separate pairs for requested and actually allowed periods. This
+could be changed later if it turns out to be necessary.
+
+### Communication with the Claims Repository
+
+When changes to the status of an access request have been made, these need to
+be communicated to the Claims Repository, which is the source of truth for all
+controlled access grants in GHGA.
+
+To allow the ARS to send changed permissions to the Claims Repository,
+the latter provides the following *internal* endpoint:
+
+- `POST /download-access/users/{user_id}/datasets/{dataset_id}`
+  - authorization: only internal from ARS (via service mesh)
+  - request body:
+    - `access_starts`: date
+    - `access_ends`: date
+  - returns nothing
+
+This will add a controlled access grant for the specified user and dataset
+with the given dates will be added to the Claims Repository.
+
+In order to facilitate authorization, the path of these endpoints starts with
+`download-access` and not with `users` which is already used by other endpoints
+of the user registry and claims repository. Also note that a corresponding
+GET endpoint already exists for use by the Download Controller Service.
+A service mesh policy should be implemented that allows the Access Request
+Service to only use the POST method, and the Download Controller Service to
+only use the GET method.
+
+### Configuration
+
+In order to know which data stewards should be notified, the configuration of
+the service should include a parameter that is a list of all email addresses
+that should be notified. This could be set to the email address of one or
+more responsible data steward(s), or just contain the email of the help-desk.
+
+The configuration should also contain parameters for the maximum time interval
+that the start date of the access request can be postponed and the maximum time
+interval between start date and end date (the maximum access validity period).
+This configuration must be added to both the Data Portal and the ARS, since it
+is used for default values and validation on both the frontend and the backend.
```

Comment:
If we don't allow changing the time right now, we might also skip the configuration of the Data Portal.

Reply:
Unless we want to show the default for the default access duration in the frontend.